player/score_bounded_mcts/score_bounded_mcts_node.py

```diff
@@ -0,0 +1,47 @@
+import numpy as np
+
+from game_logic.game_status import GameStatus
+from game_logic.igame_state import IGameState
+from player.mcts.mcts_node import MctsNode
+from player.mcts.imcts_node import IMctsNode
+
+
+class ScoreBoundedMctsNode(MctsNode):
+    def __init__(self, game_state: IGameState, action: int, parent: IMctsNode):
+        super().__init__(game_state, action, parent)
+        self._score_lower_bound = -np.inf
+        self._score_upper_bound = np.inf
+
+    def select(self, node, state):
+        while not node.untried_actions and node.game_state.available_moves:
+            node = node.select_child(self.game_state.is_first_player_move)
+            state.make_move(node.action)
+        return node, state
+    
+    def backpropagate(self, node, state):
+        current_reward = state.get_results_for_player(self.game_state.is_first_player_move)
+        while node:
+            node._number_of_runs += 1
+            is_terminal_node = state.game_status != GameStatus.InProgress
+            node._reward += current_reward
+            node._reward_squared += current_reward ** 2
+            node.update_score_bounds(current_reward, self.game_state.is_first_player_move, is_terminal_node)
+            node = node.parent
+
+    def select_child(self, is_first_player: bool):
+        if is_first_player:
+            bounded_children = [child._score_upper_bound >= self._score_lower_bound for child in self.children]
+        else:
+            bounded_children = [child._score_upper_bound >= self._score_lower_bound for child in self.children]
+        return max(bounded_children, key=lambda child: child.eval)
+
+    def update_score_bounds(self, score: float, is_first_player: bool, is_terminal: bool):
```

Comment:
Just a detail, but maybe instead of `is_first_player` we could use `is_max` or something similar?